[macOS] Add a mechanism to temporarily suppress text indicators during scrolling/resizing/zooming

pxlcoder opened the pull request:
#### d74204522d48951c10bb31282e991ce3a7df96d7
<pre>
[macOS] Add a mechanism to temporarily suppress text indicators during scrolling/resizing/zooming
<a href="https://bugs.webkit.org/show_bug.cgi?id=274071">https://bugs.webkit.org/show_bug.cgi?id=274071</a>
<a href="https://rdar.apple.com/127271298">rdar://127271298</a>

Reviewed by Abrar Rahman Protyasha and Richard Robinson.

Modern text indicators are inserted as subviews of `WKWebView`. During scrolling,
resizing, and zooming, text can get displaced from the text indicator. Consequently,
in this scenarios, text indicators should be temporarily suppressed, and restored
when in a stable state.

* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView setPageZoom:]):
(-[WKWebView _setTextZoomFactor:]):
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::scrollingNodeScrollViewDidScroll):
* Source/WebKit/UIProcess/PageClient.h:
(WebKit::PageClient::willBeginViewGesture):
(WebKit::PageClient::didEndViewGesture):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.cpp:
(WebKit::RemoteScrollingCoordinatorProxy::scrollingTreeNodeDidScroll):

Remove the platform conditional, so that all types of scrolls can be detected on macOS.

* Source/WebKit/UIProcess/ViewGestureController.cpp:
(WebKit::ViewGestureController::willBeginGesture):
(WebKit::ViewGestureController::didEndGesture):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::willBeginViewGesture):
(WebKit::WebPageProxy::didEndViewGesture):
* Source/WebKit/UIProcess/WebPageProxy.h:

Introduce `willBeginViewGesture` and `didEndViewGesture` to enable suppression
during magnification and navigation gestures.

* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::WebPageProxy::scrollingNodeScrollViewDidScroll): Deleted.
* Source/WebKit/UIProcess/mac/PageClientImplMac.h:
* Source/WebKit/UIProcess/mac/PageClientImplMac.mm:
(WebKit::PageClientImpl::scrollingNodeScrollViewDidScroll):
(WebKit::PageClientImpl::willBeginViewGesture):
(WebKit::PageClientImpl::didEndViewGesture):
* Source/WebKit/UIProcess/mac/WKTextIndicatorStyleManager.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::WebViewImpl):
(WebKit::WebViewImpl::renewGState):

Suppress modern text indicators during resizing.

(WebKit::WebViewImpl::hasContentRelativeChildViews const):
(WebKit::WebViewImpl::suppressContentRelativeChildViews):

Use a `HysteresisActivity` to drive suppression. `Remove` / `Restore` are used
to hide views for a longer period of time, such as during a magnification gesture.
Otherwise, `TemporarilyRemove` is used for affecting behaviors which have no
start/end, such as resizing.

(WebKit::WebViewImpl::contentRelativeViewsHysteresisTimerFired):
(WebKit::WebViewImpl::restoreContentRelativeChildViews):
(WebKit::WebViewImpl::setMagnification):

Canonical link: <a href="https://commits.webkit.org/278818@main">https://commits.webkit.org/278818@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d97bea0bad34f29c340d760449a19231f7b2fa84

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/51589 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/30899 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/3946 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/54856 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/2282 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/37237 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/1963 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/41994 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/53688 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/28536 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/44499 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/23122 "Build is in progress. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; run-api-tests (warnings)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/25835 "Passed tests") | | [⏳ 🛠 wpe-skia ](https://ews-build.webkit.org/#/builders/WPE-Skia-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/47797 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/1850 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/56448 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/26709 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/1717 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/49393 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/27948 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/44567 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/48591 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/28842 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7531 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/27684 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebKit/UIProcess/mac/WebViewImpl.h:
```diff
@@ -383,6 +388,9 @@ class WebViewImpl final : public CanMakeWeakPtr<WebViewImpl>, public CanMakeChec
     void setEditableElementIsFocused(bool);
 
+    enum class ContentRelativeChildViewsSuppressionType : uint8_t { Remove, Restore, TemporarilyRemove };
+    void suppressContentRelativeChildViews(ContentRelativeChildViewsSuppressionType);
```
can this use a template parameter type instead of a normal parameter (since the callsite parameter is never variable)

pxlcoder replied:
sure

pxlcoder replied:
Chatted with Richard offline – this offers minor benefit, for more typing. Keeping as-is.

pxlcoder committed:
[macOS] Add a mechanism to temporarily suppress text indicators during scrolling/resizing/zooming
https://bugs.webkit.org/show_bug.cgi?id=274071
rdar://127271298

Reviewed by Abrar Rahman Protyasha and Richard Robinson.

Modern text indicators are inserted as subviews of `WKWebView`. During scrolling,
resizing, and zooming, text can get displaced from the text indicator. Consequently,
in this scenarios, text indicators should be temporarily suppressed, and restored
when in a stable state.

* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView setPageZoom:]):
(-[WKWebView _setTextZoomFactor:]):
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::scrollingNodeScrollViewDidScroll):
* Source/WebKit/UIProcess/PageClient.h:
(WebKit::PageClient::willBeginViewGesture):
(WebKit::PageClient::didEndViewGesture):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.cpp:
(WebKit::RemoteScrollingCoordinatorProxy::scrollingTreeNodeDidScroll):

Remove the platform conditional, so that all types of scrolls can be detected on macOS.

* Source/WebKit/UIProcess/ViewGestureController.cpp:
(WebKit::ViewGestureController::willBeginGesture):
(WebKit::ViewGestureController::didEndGesture):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::willBeginViewGesture):
(WebKit::WebPageProxy::didEndViewGesture):
* Source/WebKit/UIProcess/WebPageProxy.h:

Introduce `willBeginViewGesture` and `didEndViewGesture` to enable suppression
during magnification and navigation gestures.

* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::WebPageProxy::scrollingNodeScrollViewDidScroll): Deleted.
* Source/WebKit/UIProcess/mac/PageClientImplMac.h:
* Source/WebKit/UIProcess/mac/PageClientImplMac.mm:
(WebKit::PageClientImpl::scrollingNodeScrollViewDidScroll):
(WebKit::PageClientImpl::willBeginViewGesture):
(WebKit::PageClientImpl::didEndViewGesture):
* Source/WebKit/UIProcess/mac/WKTextIndicatorStyleManager.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::WebViewImpl):
(WebKit::WebViewImpl::renewGState):

Suppress modern text indicators during resizing.

(WebKit::WebViewImpl::hasContentRelativeChildViews const):
(WebKit::WebViewImpl::suppressContentRelativeChildViews):

Use a `HysteresisActivity` to drive suppression. `Remove` / `Restore` are used
to hide views for a longer period of time, such as during a magnification gesture.
Otherwise, `TemporarilyRemove` is used for affecting behaviors which have no
start/end, such as resizing.

(WebKit::WebViewImpl::contentRelativeViewsHysteresisTimerFired):
(WebKit::WebViewImpl::restoreContentRelativeChildViews):
(WebKit::WebViewImpl::setMagnification):

Canonical link: https://commits.webkit.org/278818@main
